contracts/crashpunks-v2-friedger.clar

```diff
@@ -0,0 +1,345 @@
+        ;; 2. Mint the v2 NFT with the same nftIndex for contract-caller
+        (try! (nft-mint? crashpunks-v2 nftIndex contract-caller))
+
+        ;; 3. Burn the v1 NFT
+        (try! (contract-call? .crashpunks-v1 burn nftIndex (as-contract tx-sender)))
+        (ok true)
+    )
+)
+
+(define-public (batch-upgrade-v1-to-v2 (entries (list 200 uint)))
+    (ok (fold upgrade-v1-to-v2-helper entries false))
+)
+
+(define-public (mint-token)
+    (let (
+            (mintCounter (var-get mint-counter))
+            (mintPassBalance (get-mint-pass-balance contract-caller))
+        )
+        (asserts! (< mintCounter COLLECTION-MAX-SUPPLY) ERR-COLLECTION-LIMIT-REACHED)
+        (asserts! (> mintPassBalance u0) ERR-MINT-PASS-LIMIT-REACHED)
+
+        (try! (paymint-split MINT-PRICE contract-caller))
+        (try! (nft-mint? crashpunks-v2 mintCounter contract-caller))
+        (var-set mint-counter (+ mintCounter u1))
+        (map-set mint-pass contract-caller (- mintPassBalance u1))
+        (ok true)
+    )
+)
+
+;; only size of list matters, content of list doesn't matter
+(define-public (batch-mint-token (entries (list 20 uint)))
+    (ok (fold mint-token-helper entries false))
+)
+
+;; fail-safe: allow admin to airdrop to recipient, hopefully will never be used
+(define-public (admin-mint-airdrop (recipient principal) (nftIndex uint))
+    (begin
+        (asserts! (< nftIndex COLLECTION-MAX-SUPPLY) ERR-COLLECTION-LIMIT-REACHED)
+        (asserts! (is-eq contract-caller (var-get administrator)) ERR-NOT-ADMINISTRATOR)
+        (try! (nft-mint? crashpunks-v2 nftIndex recipient))
+        (ok true)
+    )
+)
+
+(define-public (set-mint-pass (account principal) (limit uint))
+    (begin
+        (asserts! (is-eq (var-get administrator) contract-caller) ERR-NOT-ADMINISTRATOR)
+        (ok (map-set mint-pass account limit))
+    )
+)
+
+(define-public (batch-set-mint-pass (entries (list 200 {account: principal, limit: uint})))
+    (ok (fold set-mint-pass-helper entries false))
+)
+
+;; marketplace function
+(define-public (list-item (nftIndex uint) (amount uint))
+    (begin 
+        (asserts! (unwrap! (is-approved nftIndex contract-caller) ERR-NOT-AUTHORIZED) ERR-NOT-AUTHORIZED)
+        (asserts! (> amount u0) ERR-PRICE-WAS-ZERO)
+        (ok (map-set nft-market nftIndex {price: amount}))
+    )
+)
+
+;; marketplace function
+(define-public (unlist-item (nftIndex uint))
+    (begin 
+        (asserts! (unwrap! (is-approved nftIndex contract-caller) ERR-NOT-AUTHORIZED) ERR-NOT-AUTHORIZED)
+        (ok (map-delete nft-market nftIndex))
+    )
+)
+
+;; marketplace function
+(define-public (buy-now (nftIndex uint))
+    (let 
+        (
+            (price (unwrap! (get price (map-get? nft-market nftIndex)) ERR-NFT-NOT-LISTED-FOR-SALE))
+            (owner (unwrap! (nft-get-owner? crashpunks-v2 nftIndex) ERR-COULDNT-GET-NFT-OWNER))
+            (buyer contract-caller)
+        )
+        (try! (payment-split nftIndex price contract-caller))
```

Comment:
Royalties are only paid when using `buy-now`. Other operators might not pay royalties. Is that intentional?

Maybe make the marketplace a separate contract so that all operators are equal?

How to deal with operators that have listed the nft on their contract and the nft is sold on a different one?